src/plugins/KeyboardMapping.js

```diff
@@ -0,0 +1,748 @@
+     */
+    constructor(instance, i18n) {
+        // Reference instance
+        this.instance = instance;
+        this.i18n = i18n || {};
+
+        // Register plugin
+        this.instance.KeyboardMapping = this;
+
+        this.keyboardCallbacks = [];
+
+        this.sequences = {};
+        this.requestToGenerate = [];
+        this.currentlyPressedKeys = [];
+
+        // internal state
+        this.state = new Proxy(
+            {
+                isActive: false,
+                /*
+                 * when a dialog is open we want disable listeners for keyMapping, but when dialog is closed we want to re-enable them
+                 * this is the isPaused purpose
+                 */
+                isPaused: false,
+                mappedKeysConfig: {},
+                workingMappedKeysConfig: {},
+            },
+            {
+                set: (state, prop, value) => {
+                    state[prop] = value;
+                    switch (prop) {
+                        case 'isActive':
+                            this.activatePlugin();
+                            break;
+                        case 'isPaused':
+                            if (value) {
+                                this.toolbarBtnImage.classList.add('keymapping-paused');
+                                this.state.isActive = false;
+                            } else {
+                                this.toolbarBtnImage.classList.remove('keymapping-paused');
+                                this.state.isActive = true;
+                            }
+                            break;
+                        case 'mappedKeysConfig':
+                            this.setupMappedKeysConfig();
+                            break;
+                        default:
+                            break;
+                    }
+                    return true;
+                },
+            },
+        );
+
+        // pause the listeners when dialog is open and plugin isActive
+        this.instance.store.subscribe(({overlay: {isOpen}}) => {
+            if (isOpen && this.state.isActive) {
+                this.state.isPaused = true;
+            } else if (!isOpen && this.state.isPaused) {
+                this.state.isPaused = false;
+            }
+        });
+
+        // Display widget
+        this.renderToolbarButton();
+
+        // register api function
+
+        // set config file
+        this.instance.apiManager.registerFunction({
+            name: 'setConfig',
+            category: 'keyMapping',
+            fn: (config) => {
+                // check it's a valid JSON
+                try {
+                    this.state.mappedKeysConfig = JSON.parse(JSON.stringify(config));
+                } catch (err) {
+                    throw new Error('Invalid JSON');
+                }
+                this.state.config = config;
+            },
+            description: 'Submit a config for mapping keys',
```

Comment:
maybe explain that the function expects a stringified json as arg? By the way, why not letting the user do the json parsing, that way we could directly take a js object as arg?